openhcl/openhcl_boot/src/hypercall.rs

```diff
@@ -117,15 +131,27 @@ impl HvCall {
     /// Call before jumping to kernel.
     pub fn uninitialize(&mut self) {
         if self.initialized {
-            crate::arch::hypercall::uninitialize();
             self.initialized = false;
+
+            #[cfg(target_arch = "x86_64")]
+            if self.tdx_io_page.is_some() {
+                return self.uninitialize_tdx();
+            }
+
+            crate::arch::hypercall::uninitialize();
         }
     }
 
     /// Returns the environment's VTL.
-    pub fn vtl(&self) -> Vtl {
-        assert!(self.initialized);
-        self.vtl
+    pub fn vtl(&mut self) -> Vtl {
+        self.init_if_needed();
```

Comment:
hm, should we instead go back to the previous instance where this could only be called when initialized? we can't trust the hypervisor to return the right thing in this case, so i think we need to just read after an init call